haskell.nix/.buildkite/pipeline.yml

```diff
@@ -0,0 +1,28 @@
+steps:
+
+# build haskell components
+- label: build
+  commands:
+  - nix-build -A checks.x86_64-linux.build-all
+
+# don't run tests until all components are built
+- wait
+
+# run the tests
+- label: test
+  commands:
+  - nix-build -A checks.x86_64-linux.test
+
+# run hlint
+##- label: hlint
+##  commands:
+##  - nix run -f. legacyPackages.x86_64-linux.haskellPackages.hlint
```

Comment:
In `edna` we are not using `nix run`, we are using `nix shell` instead, something like `nix-shell --run 'hlint backend -j'`. I don't know/remember the reasons and `nix run` actually looks neater to me for two reasons:
1. You can explicitly specify what you need in environment while in case of `nix shell` you have to pass everything to `devShell = pkgs.mkShell` and then each invocation of `nix-shell` builds all stuff.
2. `nix run` is just shorter than `nix-shell --run` (but ends up being longer due to explicit passing of all required environment).

IIRC the reason we are using `nix shell` is that `nix run` changed in newer Nix (≥ 3.0 or something like that) and thus `nix shell` is more future-proof? Or not? I don't know :shrug:

Reply:
You need to be clear that what's used in edna is `nix-shell` and not `nix shell`. `nix shell` is actually what `nix run` got renamed to in nix 3.0 https://github.com/NixOS/nix/commit/5d8504b9789ffebabe8226227c4061dd48354177 (and it might be renamed again later :pk:). And `nix-shell` is an older command which has a stable interface since nix 1.x.

It's somewhat a matter of taste which one to use, but to me `nix run` seems preferable too, for the first reason you mentioned (and also because `nix-shell` semantics are just weird). It's true that using `nix-shell` will help against the nix 3.0 renames, but we can deal with them when we switch to it anyway.